File: Algorithms/24. Swap Nodes in Pairs/golang/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	dummy := &ListNode{Next: head}
	prevNode := dummy

	for head != nil && head.Next != nil {
		firstNode, secondNode := head, head.Next

		prevNode.Next = secondNode
		firstNode.Next = secondNode.Next
		secondNode.Next = firstNode

		prevNode = firstNode
		head = firstNode.Next
	}

	return dummy.Next
}

File: Algorithms/24. Swap Nodes in Pairs/javascript/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function (head) {
    const dummy = new ListNode(-1, head);
    let prevNode = dummy;

    while (head && head.next) {
        let firstNode = head;
        let secondNode = head.next;

        prevNode.next = secondNode;
        firstNode.next = secondNode.next;
        secondNode.next = firstNode;

        prevNode = firstNode;
        head = firstNode.next;
    }
    return dummy.next;
};
